text/0000-more-api-documentation-conventions.md

```diff
@@ -0,0 +1,409 @@
+[using-markdown]: #using-markdown
+
+Within doc comments, use Markdown to format your documentation.
+
+Use top level headings # to indicate sections within your comment. Common headings:
+
+* Examples
+* Panics
+* Errors
+* Safety
+* Aborts
+* Undefined Behavior
+
+Even if you only include one example, use the plural form: ‘Examples’ rather
+than ‘Example’. Future tooling is easier this way.
+
+Use graves (`) to denote a code fragment within a sentence.
+
+Use triple graves (```) to write longer examples, like this:
+
+    This code does something cool.
+
+    ```rust
+    let x = foo();
+
+    x.bar();
+    ```
+
+When appropriate, make use of Rustdoc’s modifiers. Annotate triple grave blocks with
+the appropriate formatting directive.
+
+    ```rust
+    println!("Hello, world!");
+    ```
+
+    ```ruby
+    puts "Hello"
+    ```
+
+In API documentation, feel free to rely on the default being ‘rust’:
+
+    /// For example:
+    ///
+    /// ```
+    /// let x = 5;
+    /// ```
+
+In long-form documentation, always be explicit:
+
+    For example:
+
+    ```rust
+    let x = 5;
+    ```
+
+This will highlight syntax in places that do not default to ‘rust’, like GitHub.
+
+Rustdoc is able to test all Rust examples embedded inside of documentation, so
+it’s important to mark what is not Rust so your tests don’t fail.
+
+References and citation should be linked ‘reference style.’ Prefer
+
+```
+[Rust website]
+
+[Rust website]: http://www.rust-lang.org
+```
+
+to
+
+```
+[Rust website](http://www.rust-lang.org)
+```
+
+### Examples in API docs
+[examples-in-api-docs]: #examples-in-api-docs
+
+Everything should have examples. Here is an example of how to do examples:
+
+```
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, "bar");
+/// ```
+///
+/// Passing a closure to compare with, rather than a string:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, |a| a.chars().is_whitespace().all());
+/// ```
+```
+
+For particularly simple APIs, still say “Examples” and “Basic usage:” for
+consistency’s sake.
+
+### Referring to types
+[referring-to-types]: #referring-to-types
+
+When talking about a type, use its full name. In other words, if the type is generic,
+say `Option<T>`, not `Option`. An exception to this is lengthy bounds. Write `Cow<'a, B>`
+rather than `Cow<'a, B> where B: 'a + ToOwned + ?Sized`.
+
+Another possibility is to write in lower case using a more generic term. In other words,
+‘string’ can refer to a `String` or an `&str`, and ‘an option’ can be ‘an `Option<T>`’.
+
+### Link all the things
+[link-all-the-things]: #link-all-the-things
+
+A major drawback of Markdown is that it cannot automatically link types in API documentation.
+Do this yourself with the reference-style syntax, for ease of reading:
+
+```
+/// The [`String`] passed in lorum ipsum...
+///
+/// [`String`]: ../string/struct.String.html
+```
```

Comment:
I think this should be solved in rustdoc tool rather than manually. For instance, it should be trivial in a preprocessing step, assuming there's an index of symbols, to find all `[]` links that do not have a corresponding reference link and look them up in the symbol index for the proper URL.